content/blog/2022-05-23-vectorization/index.md

```diff
@@ -0,0 +1,238 @@
+
+## Vector Language
+
+The design of the vector language was based off resources from several students who had previously taken CS 6120, as well as the LLVM vector language. The students' languages are located at [vril](https://www.cs.cornell.edu/courses/cs6120/2019fa/blog/vril-vector-bril/), and [vector-instruction-support](https://www.cs.cornell.edu/courses/cs6120/2019fa/blog/interpreter-vector-support/). The LLVM vector language can be found be found at [llvm-vector](https://llvm.org/docs/LangRef.html#vector-operations).
+
+Using these resources, I specified a minimal vector language. Each vector is comprised of exactly 4 integer values. To create a fresh vector, one can use the `veczero` command, which creates a vector with all lanes starting with a value of 0. Operations on vectors include `vecadd`, `vecsub`, `vecmul`, and `vecdiv`, which are similar to their single integer Bril counterparts, except that the vector operations operate on 4 integers at a time. `vecmove` copies the value in one vector register to another vector register. `vecmac` represents an operation multiplying two vectors, and then adding a vector offset. `vecneg` negates all the elements of a vector. Due to project limitations, neither `vecmac` nor `vecneg` were handled; however, it is possible to extend the vectorization to handle these 2 instructions. 
+
+To interact between vector registers and normal registers. one can load into and store from vectors. For instance, the `vecload` instruction places one integer of data into a vector, at the specified index in the vector. Likewise, `vecstore` extracts one integer of data from a vector into a Bril pseudo-register. 
+
+The vector language is simulated in the Bril interpreter. I implemented the vector registers as variables storing array of width 4. Each vector operation manipulates the array appropriately; for instance, `veczero` instantiates all the cells of the array to have value 0, `vecload v i d` loads data `d` into the array representing vector `v` at index `i` in the array, and `vecadd v1 v2` does element-wise addition on the arrays representing vectors `v1` and `v2`. The implementation of this interpreter is located in the files `bril-vc` and `brili-vc`. I also augment the `print` instruction to print out all elements in a vector, for easy debugging.
```

Comment:
Very nice! I like this design for a vector extension quite a bit. Would you be interested in contributing some aspects of it to the Bril monorepo so other projects could conceivably build on it in the future?

The one design aspect that I could imagine changing is adding "whole-vector" loads and stores that interact with the memory extension. That is, `v: vec = vecld p` would get 4 integers from memory starting at pointer `p` and put it into vector register `v`. (One could imagine renaming the existing `vecload` and `vecstore` to evoke "insert" and "extract".) Having vectorized loads/stores would bring us even closer to modeling the performance advantage of real vector ISAs.

Reply:
I would be interested in contributing to the Bril monorepo! I agree that the "whole-vector" loads would be nice, and that the renaming would be more consistent.